src/axom/quest/InOutOctree.hpp

```diff
@@ -60,3362 +65,1516 @@ namespace quest
+   *
+   * \pre This function assumes the vertices have been inserted and the mesh has been reordered
+   * \param tIdx The index of the cell to check
+   * \param blk The block that we are checking against
+   * \return true if one of the cell's vertices are indexed by \a blk, false otherwise
    */
-  friend bool operator==(const DynamicGrayBlockData& lhs,
-                         const DynamicGrayBlockData& rhs)
+  bool blockIndexesElementVertex(CellIndex tIdx, const BlockIndex& blk) const
   {
-    return
-      //(static_cast<const BlockData&>(lhs) == static_cast<const
-      // BlockData&>(rhs))
-      //&&
-      (lhs.m_vertIndex == rhs.m_vertIndex) &&
-      (lhs.m_tris.size() == rhs.m_tris.size())  // Note: We are not
-                                                // checking the contents
-      // && (lhs.m_tris == rhs.m_tris)                //       of the triangle
-      // array, only the size
-      && lhs.m_isLeaf == rhs.m_isLeaf;
-  }
-
-public:  // Functions related to whether this is a leaf
-  /** Predicate to determine if the associated block is a leaf in the octree */
-  bool isLeaf() const { return m_isLeaf; }
+    SLIC_ASSERT(m_generationState >= INOUTOCTREE_MESH_REORDERED);
 
-  /** Sets a flag to determine whether the associated block is a leaf or
-     internal */
-  void setLeafFlag(bool isLeaf) { m_isLeaf = isLeaf; }
+    CellVertIndices tVerts = m_meshWrapper.cellVertexIndices(tIdx);
+    for(int i = 0; i < tVerts.size(); ++i)
+    {
+      // Using the vertex-to-block cache to avoid numerical degeneracies
+      if(blockIndexesVertex(tVerts[i], blk)) return true;
+    }
+    return false;
+  }
 
-public:  // Functions related to the associated vertex
   /**
-   * \brief Checks whether there is a vertex associated with this leaf
+   * \brief Determines whether the specified 3D point is within the gray leaf
+   *
+   * \param queryPt The point we are querying
+   * \param leafBlk The block of the gray leaf
+   * \param data The data associated with the leaf block
+   * \return True, if the point is inside the local surface associated with this
+   * block, false otherwise
    */
-  bool hasVertex() const { return m_vertIndex >= 0; }
-
-  /** Sets the vertex associated with this leaf */
-  void setVertex(VertexIndex vInd) { m_vertIndex = vInd; }
-
-  /** Clears the associated vertex index */
-  void clearVertex() { m_vertIndex = NO_VERTEX; }
-
-  /** Accessor for the index of the vertex associated with this leaf */
-  VertexIndex& vertexIndex() { return m_vertIndex; }
-
-  /** Constant accessor for the index of the vertex associated with this leaf */
-  const VertexIndex& vertexIndex() const { return m_vertIndex; }
-
-public:  // Functions related to the associated triangles
-  /** Check whether this Leaf has any associated triangles */
-  bool hasTriangles() const { return !m_tris.empty(); }
+  template <int TDIM>
+  typename std::enable_if<TDIM == 3, bool>::type withinGrayBlock(
```

Comment:
Is the goal of the `enable_if` stuff to have different implementations for the 2D and 3D case?

If so, I think you could also achieve this with explicit member function specializations instead of SFINAE, e.g.:

```c++

template<int DIM>
class InOutOctree
{
   //...
   bool withinGrayBlock(params...) const;
};

template<>
bool InOutOctree<3>::withinGrayBlock(params...) const
{
   // 3D code here...
}

template<>
bool InOutOctree<2>::withinGrayBlock(params...) const
{
   // 2D code here...
}
```

Then you'd only need one declaration of `withinGrayBlock` in the class.

Reply:
Thanks @kanye-quest!

I tried something like that originally, but didn't get it to work.
Your suggestion worked and I updated the code.